trees/utils/RobustOCT.py

```diff
@@ -0,0 +1,147 @@
+import copy
+from gurobipy import Model, LinExpr, GRB, quicksum
+import math
+import numpy as np
+import pandas as pd
+import sys
+
+class RobustOCT:
+    def __init__(self, X, y, tree, X_col_labels, labels, time_limit, costs, budget):
+        '''
+        :param data: The training data
+        :param label: Name of the column representing the class label
+        :param tree: Tree object
+        :param time_limit: The given time limit for solving the MIP
+        :param costs: The costs of uncertainty
+        :param budget: The budget of uncertainty
+        '''
+        self.cat_features = X_col_labels
+        self.X = X
+        self.y = y
+        self.labels = labels        
+        
+        # datapoints contains the indicies of our training data
+        self.datapoints = np.arange(0, self.X.shape[0])
+        self.tree = tree
+
+        # Regularization term: encourage less branching without sacrificing accuracy
+        self.reg = 1 / (len(tree.Nodes) + 1)
+
+        '''
+        Get range of data, and store indices of branching variables based on range
+        '''
+        min_values = X.min(axis=0)
+        max_values = X.max(axis=0)
+        f_theta_indices = []
+        b_indices = []
+        for f in self.cat_features:
+            min_value = min_values[f]
+            max_value = max_values[f]
+
+            # cutoffs are from min_value to max_value - 1
+            for theta in range(min_value, max_value): 
+                f_theta_indices += [(f,theta)]
+                b_indices += [(n, f, theta) for n in self.tree.Nodes]
+
+        self.min_values = X.min(axis=0)
+        self.max_values = X.max(axis=0)
+        self.f_theta_indices = f_theta_indices
+        self.b_indices = b_indices
+
+        '''
+        Create uncertainty set
+        '''
+        self.epsilon = budget # Budget of uncertainty
+        self.gammas = costs # Cost of feature uncertainty
+        self.eta = budget + 1 # Cost of label uncertainty - future work
+
+        # Decision Variables
+        self.t = 0
+        self.b = 0
+        self.w = 0
+
+        # Gurobi model
+        self.model = Model('RobustOCT')
+        # The cuts we add in the callback function would be treated as lazy constraints
+        self.model.params.LazyConstraints = 1
+
+        '''
+        To compare all approaches in a fair setting we limit the solver to use only one thread to merely evaluate 
+        the strength of the formulation.
+        '''
+        # self.model.params.Threads = 1
+        self.model.params.TimeLimit = time_limit
+        
+        # NOTE - Should we allow users to suppress logging?
+        # self.model.params.OutputFlag = 0 # supress all logging
```

Comment:
Yes, I think we should give users this option. Maybe with a verbose argument for when they initialize the class